### PR TITLE
Feature stake lock (step1)

### PR DESCRIPTION
**Pull request by AndriiAndrusyshyn**
## Consensus Changes
- [x] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 1]
```diff
@@ -28,6 +29,7 @@ namespace eosiosystem {
 
    static constexpr uint32_t refund_delay_sec = 3*24*3600;
    static constexpr int64_t  ram_gift_bytes = 1400;
+   const eosio::time_point two_month = eosio::time_point(eosio::seconds(5097600)); // two month
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: @AndreiAndrusyshyn I'm not yet sure about best way for non-constexpr constants but it seems better to keep it static and initialize in cpp;

as well you can use eosio::days it's more readable for human

two_month is unused var ( I suppose it should be 6month); anyway its to specific name, something like resignation_period/stake_lock_period is better

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 2]
```diff
@@ -449,15 +451,26 @@ namespace eosiosystem {
...
-      check( unstake_cpu_quantity >= zero_asset, "must unstake a positive amount" );
-      check( unstake_net_quantity >= zero_asset, "must unstake a positive amount" );
-      check( unstake_cpu_quantity.amount + unstake_net_quantity.amount > 0, "must unstake a positive amount" );
-      check( _gstate.total_activated_stake >= min_activated_stake,
-             "cannot undelegate bandwidth until the chain is activated (at least 15% of all tokens participate in voting)" );
-
-      changebw( from, receiver, -unstake_net_quantity, -unstake_cpu_quantity, false);
-   } // undelegatebw
+
+        if(is_block_producer(from))
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn It's better to place this validation after EOS checks so we will have less diff

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 3]
```diff
@@ -449,15 +451,26 @@ namespace eosiosystem {
...
+        {
+            del_bandwidth_table     del_tbl( _self, from.value );
+            auto itr = del_tbl.find( from.value );
+
+            del_tbl.modify(itr, from, [&]( auto& dbo ){
+                check(current_time_point() > dbo.staked_time + eosio::days(180), "producer cannot undelegate bandwidth during six month" );
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn
I suppose something like this should work:
```
check( itr == del_tbl.end(), "user has no resources" ); // not sure if this the case

const auto err = "producer cannot undelegate bandwidth during"s + std::to_string( resignation_period_in_days ) + " days"s;
check( current_time_point() > itr->staked_time, err );
```

File: contracts/contracts/eosio.system/src/eosio.system.cpp [hunk 4]
```diff
@@ -16,7 +16,7 @@
 namespace eosiosystem {
 
    const int64_t  inflation_precision           = 100;     // 2 decimals
-   const int64_t  default_annual_rate           = 500;     // 5% annual rate
+   const int64_t  default_annual_rate           = 500;     // 5%  annual rate
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/eosio.system/src/eosio.system.cpp]: @AndriiAndrusyshyn please revert any unnecessary changes

File: unittests/bootseq_tests.cpp [hunk 5]
```diff
@@ -370,4 +369,101 @@ BOOST_FIXTURE_TEST_CASE( bootseq_test, bootseq_tester ) {
...
+            );
+        };
+
+        votepro( N(b1), { N(proda)} );
+        votepro( N(whale2), {N(proda)} );
+        votepro( N(whale3), {N(proda)} );
+        votepro( N(whale4), {N(proda)} );
+
+        // No producers will be set, since the total activated stake is less than 150,000,000
+        produce_blocks_for_n_rounds(2); // 2 rounds since new producer schedule is set when the first block of next round is irreversible
```
Comment (mark-tsyrulnyk) [inline unittests/bootseq_tests.cpp]: @AndriiAndrusyshyn not sure but seems that we can safely delete those lines
```
        // Since the total vote stake is more than 150,000,000, the new producer set will be set
        produce_blocks_for_n_rounds(2); // 2 rounds since new producer schedule is set when the first block of next round is irreversible
        active_schedule = control->head_block_state()->active_schedule;
```

File: unittests/bootseq_tests.cpp [hunk 6]
```diff
@@ -370,4 +369,101 @@ BOOST_FIXTURE_TEST_CASE( bootseq_test, bootseq_tester ) {
...
+        //Try to unstake tokens from produceir should throw error, because unstaked condition does not met ( producer can unstake only after two month )
+        BOOST_REQUIRE_THROW(undelegate_bandwidth(N(proda), N(proda), core_from_string("1.0000"), core_from_string("1.0000")), eosio_assert_message_exception);
+
+        // Skip 6 month
+        produce_block(second_july_2020 - control->head_block_time().time_since_epoch());
+
+        // Block one should be able to unstake all his stake now
+        BOOST_REQUIRE(undelegate_bandwidth(N(proda), N(proda), core_from_string("1.0000"), core_from_string("1.0000")));
+
+        return;
```
Comment (mark-tsyrulnyk) [inline unittests/bootseq_tests.cpp]: @AndriiAndrusyshyn please remove return and everything that goes after

File: unittests/main.cpp [hunk 7]
```diff
@@ -33,8 +33,8 @@ boost::unit_test::test_suite* init_unit_test_suite(int argc, char* argv[]) {
    if(is_verbose) {
       fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::debug);
    } else {
-      fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::off);
-   }
+      fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::debug);
```
Comment (mark-tsyrulnyk) [inline unittests/main.cpp]: @AndriiAndrusyshyn I'm not 100% sure, but in latest develop such changes already done

**Review by mark-tsyrulnyk — APPROVED** [review]

File: contracts/contracts/eosio.system/include/eosio.system/eosio.system.hpp [hunk 8]
```diff
@@ -254,7 +254,7 @@ namespace eosiosystem {
        */
       double              last_vote_weight = 0; /// the vote weight cast the last time the vote was updated
       time_point          vote_mature_time;
-      
+
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/include/eosio.system/eosio.system.hpp:257]: Maybe revert this changes? File included in PR only because of spaces seems weird

File: contracts/contracts/eosio.system/include/eosio.system/eosio.system.hpp [hunk 9]
```diff
@@ -1300,7 +1300,7 @@ namespace eosiosystem {
                                             double additional_shares_delta = 0.0, double shares_rate_delta = 0.0 );
          bool does_satisfy_stake_requirement( const name& producer ) const;
          bool is_block_producer( const name& producer ) const;
-         
+
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/include/eosio.system/eosio.system.hpp:1303]: Maybe revert this changes? File included in PR only because of spaces seems weird

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 10]
```diff
@@ -295,6 +297,16 @@ namespace eosiosystem {
       check( _gstate.total_activated_stake >= min_activated_stake,
              "cannot undelegate bandwidth until the chain is activated (at least 15% of all tokens participate in voting)" );
 
+      if(is_block_producer(from))
+      {
+          del_bandwidth_table     del_tbl( _self, from.value );
+          auto itr = del_tbl.find( from.value );
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp]: You can do this with one line. Smth like this:
const auto& del = del_tbl.get( from.value, "user has no resources" );

File: contracts/contracts/eosio.system/src/voting.cpp [hunk 11]
```diff
@@ -184,7 +184,7 @@ namespace eosiosystem {
       require_auth( voter_name );
 
       check( is_block_producer( voter_name ), "only block producers are allowed to vote" );
-      
+
```
Comment (gBaGu) [inline contracts/contracts/eosio.system/src/voting.cpp:187]: Again, file is included in PR only because of removed spaces

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp [hunk 12]
```diff
@@ -295,6 +297,13 @@ namespace eosiosystem {
       check( _gstate.total_activated_stake >= min_activated_stake,
              "cannot undelegate bandwidth until the chain is activated (at least 15% of all tokens participate in voting)" );
 
+      if(is_block_producer(from))
+      {
+          del_bandwidth_table     del_tbl( _self, from.value );
+          const auto& del = del_tbl.get( from.value, "user has no resources" );
+          check( current_time_point() > del.staked_time + stake_lock_period, "producer cannot undelegate bandwidth during 180 days");
```
Comment (roman-tik) [inline contracts/contracts/eosio.system/src/delegate_bandwidth.cpp:304]: In case a BP stakes more tokens during the initial locking period, it will increase the locking proportionally. For example, an account opens as BP with 100M tokens staked with the initial six month locking period. After 1.5 months they stake an additional 200M tokens. As a result, the locking period is adjusted from the remaining 4.5 months to 5.5 months with a total of 300M tokens staked. 6mon - 6mon*(100M*1.5/6)/(100M+200M) = 5.5mon.

**Review by roman-tik — CHANGES_REQUESTED** [review]

**Review by me-vadim — APPROVED** [review]

File: contracts/contracts/eosio.system/src/delegate_bandwidth.cpp (hunk 12 above)
Comment (mark-tsyrulnyk) [reply contracts/contracts/eosio.system/src/delegate_bandwidth.cpp:304]: @roman-tik we will implement this logic but in scope of other PR

**Review by roman-tik — APPROVED** [review]
